### PR TITLE
#102 - Added LoggingSlice

### PR DESCRIPTION
**Pull request by olegmoz**
Part of #102 
Added Slice with logging, that could be composed with other Slices

### CODE REVIEW

**Comment by paulodamaso** [comment]
@g4s8 Please assign me

**Comment by 0crat** [comment]
This pull request [#103](https://github.com/artipie/http/issues/103) is assigned to @vatavuk[/z](https://www.0crat.com/u/vatavuk), here is [why](https://www.0crat.com/footprint/CT2E6TK9B/88b445aa-7424-4344-91e5-2942bae9c989); the budget is 15 minutes, see [§4](http://www.zerocracy.com/policy.html#4); please, read [§27](http://www.zerocracy.com/policy.html#27) and when you decide to accept the changes, inform @g4s8[/z](https://www.0crat.com/u/g4s8) (the architect) right in this ticket; if you decide that this PR should not be accepted ever, also inform the architect; this [blog post](http://www.yegor256.com/2015/02/09/serious-code-reviewer.html) will help you understand what is expected from a code reviewer; there will be a monetary reward for this job

File: src/test/java/com/artipie/http/slice/LoggingSliceTest.java [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+
+/**
+ * Tests for {@link LoggingSlice}.
+ *
+ * @since 0.8
+ */
+class LoggingSliceTest {
+
+    @Test
+    void shouldLogRequestAndResponse() {
```
Comment (Vatavuk) [inline src/test/java/com/artipie/http/slice/LoggingSliceTest.java:44]: @olegmoz Can we test what's actually being logged through jcabi logger? If not maybe we can consider creating a wrapper over Jcabi and inject it in the ctor...

File: src/main/java/com/artipie/http/slice/LoggingSlice.java [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+import com.artipie.http.rs.RsStatus;
+import com.jcabi.log.Logger;
+import java.nio.ByteBuffer;
+import java.util.Map;
+import java.util.concurrent.CompletionStage;
+import java.util.logging.Level;
+import org.reactivestreams.Publisher;
+
+/**
+ * Slice logging incoming requests and outgoing responses.
```
[the remark this answers is not in the document]
Comment (olegmoz) [reply src/main/java/com/artipie/http/slice/LoggingSlice.java]: @Vatavuk fixed

File: src/main/java/com/artipie/http/slice/LoggingSlice.java [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+    @Override
+    public Response response(
+        final String line,
+        final Iterable<Map.Entry<String, String>> headers,
+        final Publisher<ByteBuffer> body
+    ) {
+        final StringBuilder msg = new StringBuilder();
+        msg.append(">> ").append(line);
+        LoggingSlice.append(msg, headers);
+        final Object source = this;
```
[the remark this answers is not in the document]
Comment (olegmoz) [reply src/main/java/com/artipie/http/slice/LoggingSlice.java]: @Vatavuk done, thanks

File: src/test/java/com/artipie/http/slice/LoggingSliceTest.java (hunk 1 above)
Comment (olegmoz) [reply src/test/java/com/artipie/http/slice/LoggingSliceTest.java:44]: @Vatavuk it also possible to capture logs by inserting log4j or logback back-end for jcabi and capturing logs there, but to me it looks like an overkill

Comment (Vatavuk) [reply src/test/java/com/artipie/http/slice/LoggingSliceTest.java:44]: @olegmoz This test is useless by itself, I would assert the log printout if it is not a big hustle.

Comment (olegmoz) [reply src/test/java/com/artipie/http/slice/LoggingSliceTest.java:44]: @Vatavuk this test can be used to just run and see the results which might be convenient for making changes. Checking logging output is not easy and looks like overtesting, which is not good.

Comment (Vatavuk) [reply src/test/java/com/artipie/http/slice/LoggingSliceTest.java:44]: @olegmoz ok

**Comment by 0crat** [comment]
@vatavuk[/z](https://www.0crat.com/u/vatavuk) this job was assigned to you 5days ago. It will be taken away from you soon, unless you close it, see [§8](http://www.zerocracy.com/policy.html#8). Read [this](http://www.yegor256.com/2014/04/13/no-obligations-principle.html) and [this](http://www.yegor256.com/2014/11/24/principles-of-bug-tracking.html), please.

**Comment by Vatavuk** [comment]
@g4s8 please remove this issue from the scope, it is still in my agenda.

**Comment by g4s8** [comment]
@0crat out

**Comment by 0crat** [comment]
@sereshqua[/z](https://www.0crat.com/u/sereshqua) please review this job completed by @vatavuk[/z](https://www.0crat.com/u/vatavuk), as in [§30](http://www.zerocracy.com/policy.html#30); the job will be fully closed and all payments will be made when the quality review is completed

**Comment by sereshqua** [comment]
@Vatavuk please make sure that during next CR you will find mostly design problems, but not rather cosmetic issues

**Comment by Vatavuk** [comment]
@sereshqua ok

**Comment by sereshqua** [comment]
@0crat quality acceptable

**Pull request merged**